shared/health/src/main/java/io/pravega/shared/health/impl/ContributorRegistryImpl.java

```diff
@@ -0,0 +1,321 @@
+    synchronized public HealthContributor register(HealthComponent component, HealthComponent parent) {
+        parent = Objects.isNull(parent) ? root : parent;
+        // A HealthComponent should only exist if defined during construction, instead of adding it dynamically.
+        if (!components.contains(parent.getName()) || !contributors.containsKey(parent.getName())) {
+            log.warn("Attempting to register {} under unrecognized {} -- aborting.", component, parent);
+            return null;
+        }
+        components.add(component.getName());
+        return register(component, parent.getName());
+    }
+
+    /**
+     * Registers a {@link HealthContributor} under the specified parent {@link HealthComponent}.
+     *
+     * @param contributor The {@link HealthContributor} object to add to the registry as a child of 'parent'.
+     * @param parent The {@link HealthComponent} to add 'contributor' as a child/dependency.
+     * @return The {@link HealthContributor} registered under the {@link ContributorRegistry}.
+     */
+    @NonNull
+    synchronized public HealthContributor register(HealthContributor contributor, HealthComponent parent) {
+        parent = Objects.isNull(parent) ? root : parent;
+        return register(contributor, parent.getName());
+    }
+
+    /**
+     * Registers a {@link HealthContributor} under the specified parent {@link HealthComponent}.
+     *
+     * @param contributor The {@link HealthContributor} object to add to the registry as a child of 'parent'.
+     * @param componentName The {@link String} id of a {@link HealthComponent} to add 'contributor' as a child/dependency.
+     * @return The {@link HealthContributor} registered under the {@link ContributorRegistry}. Null return values are permitted.
+     */
+    @NonNull
+    @Override
+    synchronized public HealthContributor register(HealthContributor contributor, String componentName) {
+        // HealthContributor mapped by 'parent' should exist at time of some child registration.
+        if (!contributors.containsKey(componentName)) {
+            log.debug("Unrecognized HealthContributor::{} -- aborting registration.", componentName);
+            return null;
+        }
+        String contributorName = contributor.getName();
+        if (contributors.containsKey(contributorName)) {
+            log.warn("{} has already been registered.", contributors.get(contributorName));
+            return contributors.get(contributorName);
+        }
+        log.debug("Registering {} to {}.", contributor, contributors.get(componentName));
+        contributors.put(contributorName, contributor);
+        // A new contributor should have no child relations to overwrite.
+        // Realistically, this should be restricted because it implies non HealthComponent contributors can have children.
+        children.put(contributorName, new HashSet<>());
+        // Add the child relation.
+        children.get(componentName).add(contributor);
```

Comment:
Combine this and the above line into 1. Create a map, add something to it, then put the map in the children collection.

Reply:
I'm not sure I understand. First on 169 the children set for the soon to be registered `HealthContributor` is initialized, then on 171 the contributor is added as a child to the parent (`componentName`).